docs/source/markdown/options/add-host.md

```diff
@@ -2,10 +2,14 @@
 ####>   podman build, create, farm build, pod create, run
 ####> If file is edited, make sure the changes
 ####> are applicable to all of those.
-#### **--add-host**=*host:ip*
+#### **--add-host**=*hostname[,hostname[,…]]*:*ip* | *hostname[,hostname[,…]]*:*host-gateway*
```

Comment:
I don't think the host-gateway need to specified here, makes things more confusing IMO. If you think it is important it should properly be `...=*hostname[;hostname[;…]]*:*<ip>|host-gateway*`

Reply:
I was thinking about a similar syntax, too. However, I was unsure and thus checked other Podman options and couldn't find any other Podman option using a simliar syntax (doesn't mean that I might have missed it though). Personally I'd write it as `*hostname[;hostname[;…]]*:*ip*|"host-gateway"` to indicate that `ip` is a param and `"host-gateway"` a fixed string (I also know `<>` to indicate required params, but that isn't used in Podman either?).

Keeping this consistent is more important though. The problem is that I'm not entirely sure what the most consistent solution is :sweat_smile: I'm fine with any variant, just tell me which variant to use, I'll add it then.

About "host-gateway": I feel like that it should be mentioned in the format line, because I believe that the format line should include all possibilities of the option and allow the user to fully understand the option without reading the description. But I'm fine with not mentioning it, too. Basically I'm fine with anything :smile: I've removed it for now.

Reply:
I am not to sure either, yes the syntax is not very consistent across all our man pages.
If you like to keep host-gateway in the format line that is fine for me.

Although I guess the description could also need some more explaining what `host-gateway` is, i.e. a ip address that can be used to connect from the container to the host.

Reply:
What do you think about the suggested `*hostname[;hostname[;…]]*:*ip*|"host-gateway"`? If that syntax is okay, I'd suggest using that.

---

Not sure how to further elaborate on `host-gateway` though. Right now we have this:

> Instead of an IP address the special flag *host-gateway* can be given that resolves to an IP address the container can use to connect to the host. Podman will automatically add the `host.containers.internal` and `host.docker.internal` hostnames using the *host-gateway* address, unless the **--no-hosts** option is given.

We could add more details about how Podman chooses that IP address, but this fundamentally depends on the network mode used and even some other options as well, like if IPv4 is disabled (I assume that *host-gateway* resolves to an IPv6 address then?). I'm not entirely sure whether such details add much clarification.

I just checked Docker's docs on this, see https://docs.docker.com/reference/cli/docker/container/run/#add-host. They describe the option as follows:

> The --add-host flag supports a special host-gateway value that resolves to the internal IP address of the host. This is useful when you want containers to connect to services running on the host machine.
> 
> It's conventional to use host.docker.internal as the hostname referring to host-gateway. Docker Desktop automatically resolves this hostname, see Explore networking features.

I feel like this is all mentioned already. Suggestions?

Reply:
> What do you think about the suggested *hostname[;hostname[;…]]*:*ip*|"host-gateway"? If that syntax is okay, I'd suggest using that.

Works for me.

---

For the other point it is rather complicated, first of all there is no guarantee that `host.containers.internal` will be added. Depending on your host setup we might not be able to find a free host ip for this so I feel like it is important to mention that the entry might not be set.
With podman machine it is even more complciated as we do not add it to /etc/hosts and let the gvproxy dns resolver resolve the real host ip as podman would use the VM ip otherwise.

And this should likely point to `host_containers_internal_ip` option from containers.conf because users can overwrite the behavior there or set a fixed ip if they like.

So overall I think something like this:
```
Unless disabled in via *host_containers_internal_ip* option  in *containers.conf* Podman will add the `host.containers.internal` and `host.docker.internal` name by default with an ip pointing to the host. If Podman cannot figure out a suitable ip address it may not add this entry. In order to use a different name it is possible to use `--add-host <name>:host-gateway` and then it will use the same ip in place of host-gateway automatically, however if it failed to find a suitable ip in that case Podman will return an error instead and not silently skip the entry.
```

Reply:
Got it. See updated suggestion.

I skipped mentioning that one can force Podman to add the internal hostnames using `--add-host`. I rather added a general note that one can use `--add-host` to overwrite any automatically added entries (I actually use that for `--hostname`), therefore implying that this is also true for these internal hostnames, even though not specifically mentioning it because I assume this is very rarely used for that.

Since this yields some cross references I decided to not just update `--add-host`, but `--no-hosts`, `--hostname`, and `--name` as well. This should unify them and hopefully constitute a definitive documentation of Podman's `/etc/hosts` file. Did I miss something?

I assumed that *host_containers_internal_ip* also overwrites *host-gateway* of `--add-host`, not just the internal hostnames. Is that correct?

Open question: I checked the docs of the `host_containers_internal_ip` config and one can apparently set it to 'none' to disable adding the internal hostnames by default. Does setting `host_containers_internal_ip="none"` also affect *host-gateway* of `--add-host`? If yes, how? I assumed that it should have no effect, but is that actually true?

Reply:
> I assumed that host_containers_internal_ip also overwrites host-gateway of --add-host, not just the internal hostnames. Is that correct?

Yes

> Does setting host_containers_internal_ip="none" also affect host-gateway of --add-host? If yes, how? I assumed that it should have no effect, but is that actually true?

It will cause an error and not work but this is more of a how the code works today and I am not opposed to changing this in the future if anyone needs that.